File: Algorithms/Sorting-Algorithms/Bubble-Sort/bubble-sort.py
# Bubble Sort: Compare adjacent elements and swap. At the end of each iteration the largest element is stored at the end of the array. 
# Sorts from right to left, ie. from maximum to minimum, right side of array gets sorted first. 
# Lagest element "BUBBLES OUT" to the top/end of the array.

# TC: Worst Case / Avg Case: O(n**2)
# Best Case (already sorted array): O(n) -> only needs to traverse the array once, if no swap then break the loop and return.


# In Python, integers are immutable, so swapping two integers using a function won't have any effect outside the function. 
# You need to swap elements of the list directly.
def swap(arr: list[int], i: int, j: int) -> None:
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp


def bubble_sort(arr: list[int]) -> None:
    n = len(arr)
    for i in range(1, n):
        for j in range(n-i):
            if arr[j] > arr[j+1]:
                swap(arr, j, j+1)

    
def bubble_sort_opti(arr: list[int]) -> None:
    n = len(arr)
    didSwap = False
    for i in range(1, n):
        for j in range(n-i):
            if arr[j] > arr[j+1]:
                swap(arr, j, j+1)
                didSwap = True
        if not didSwap:
            break
   
    
def create_arr() -> list[int]:
    n = int(input("Enter the number of elements in the array: "))
    arr = []
    print("Enter the elements of array: ")
    for i in range(n):
        arr.append(int(input()))
    print()
    print("The array is: ")
    for i in arr:
        print(i, end=" ")
    print()
    return arr
    
    
if __name__ == "__main__":
    arr = create_arr()
    bubble_sort_opti(arr)
    print("Sorted Array: ")
    for i in arr:
        print(i, end=" ")
    print()

File: Algorithms/Sorting-Algorithms/Bubble-Sort/bubble-sort.cpp
#include<iostream>
#include<vector>
using namespace std;

void swap1(int &a, int &b) {
    int temp = a;
    a = b;
    b = temp;
}


void bubble_sort_vec(vector<int> &arr) {
    int n = arr.size();
    for(int i=1; i<n; i++) {
        for(int j=0; j<n-i; j++) {
            if(arr[j] > arr[j+1])
                swap1(arr[j], arr[j+1]);
        }
    }
}


void bubble_sort_opti(vector<int> &arr) {
    int n = arr.size();
    bool swap = false;
    for(int i=1; i<n; i++) {
        for(int j=0; j<n-i; j++) {
            if(arr[j] > arr[j+1]) {
                swap1(arr[j], arr[j+1]);
                swap = true;
            }
        }
        if(!swap)
            break;
    }
}


vector<int> form_vec() {
    int n;
    cout<<"Enter the number of elements of the array: ";
    cin >> n;
    vector<int> arr(n);
    cout<<"Enter the elements of the array: "<<endl;
    for(int i=0; i<n; i++)
        cin >> arr[i];
    cout<<endl;
    cout<<"The array is: "<<endl;
    for(int i: arr)
        cout<< i <<" ";
    cout<<endl<<endl;
    return arr;
}


int main() {
    vector<int> arr = form_vec();
    bubble_sort_opti(arr);
    cout<<"Sorted Array: "<<endl;
    for(int i: arr)
        cout<< i <<" ";
    cout<<endl;
    return 0;

}
